### PR TITLE
ux: add a new expandStorage handler

### PR DESCRIPTION
**Pull request by parth-gr**
update the ux backend server to use new expandstorage handler

For end to end UI testing image used for backend server: http://quay.io/paarora/rook-ceph:v637

UI will send the form 
```
type AttachStorageForm = {
  storageClassForOSDs: string
  enableEncryption: boolean
  storage: "1"
  replica: int
  count: int
  strogaeClusterName: string
  poolDetails: { 
    volumeType: 'Filesystem' | 'Block'
    poolName: string
    dataProtectionPolicy: '2Way' | '3Way'
    enableCompression: boolean
    filesystemName: string
    failureDomain: string 
  }
  storageClassDetails: {
    reclaimPolicy: 'Retain' | 'Delete'
    name: string 
    volumeBindingMode: 'Immediate' | 'WaitForFirstConsumer'
    enableStorageClassEncryption : bool
    encryptionKMSID: string
  }
}
```

### CODE REVIEW

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (rchikatw) [inline services/ux-backend/handlers/expandstorage/handler.go:91]: you have missed adding the rbac as well you can add it here: https://github.com/red-hat-storage/ocs-operator/tree/main/config/rbac

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:66]: Will update

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:58]: I can use the same client function, once that PR get merged, 
No need to create two clients,

Will open a future re-factor PR

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:53]: So far from UI it is mandatory to use KMS with StorageClass encryption. Here I can see `EnableStorageClassEncryption` is of type `bool`, so we are certain that we don't want KMS here ??

Will Ceph-CSI use K8s Secret instead to store keys ??

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:134]:
```suggestion
	compression := "none"
```

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:145]: also need to set `enableCrushUpdates: true,` spec...

Ref:
https://github.com/rook/rook/pull/14322
https://github.com/red-hat-storage/ocs-operator/pull/2751

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:205]:
```suggestion
	compression := "none"
```

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:145]: will FD always be "host", just to confirm that we are sure ??

asking because UI reads it from the StorageCluster CR's status and set the value accordingly, instead of hardcoding "host" always.

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:221]: same FD hardcoding related question here as well...

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [inline services/ux-backend/handlers/expandstorage/handler.go:261]: just want to point out that UI doesn't support encryption for CephFS prov classes yet... I hope Eran is aware that this feature does not follow existing behaviour, if not then please bring this up with him once.

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:53]: We will use KMS itself the cluster-wide settings,
This field `EnableStorageClassEncryption` can set per device set encryption, whether a device set needs encryption or not,

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:145]: We can get from UI then

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [reply services/ux-backend/handlers/expandstorage/handler.go:54]: IIUC field `enableEncryption` is for deviceSet encryption and here we could use cluster-wide details.

but I am confirming about `enableStorageClassEncryption` field, which is for StorageClass encryption (not OSD/deviceSet), Ceph-CSI uses different ConfigMap than Rook for KMS.

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (malayparida2000) [reply services/ux-backend/handlers/expandstorage/handler.go]: Yes, we will need that fiag in a new block pool that is created. Sanjal has referenced the correct PRs, There was a UI change as well to add this option to any ned block pool getting created via UI

Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go]: Added

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:54]: I am specifying this using bool setting  https://github.com/rook/rook/blob/master/deploy/examples/csi/rbd/storageclass.yaml#L51

can UI also provide https://github.com/rook/rook/blob/master/deploy/examples/csi/rbd/storageclass.yaml#L56?

Comment (SanjalKatiyar) [reply services/ux-backend/handlers/expandstorage/handler.go:54]: We need to display the same KMS form for this feature that UI currently displays on "Create StorageClass" page.

After that it's a design question that who will create/patch the csi-kms ConfigMap with KMS details. On StorageClass creation page (day-2) and during initial deployment (day-1) UI creates it... but here we are introducing new API endpoint, so whether this API should do it now or UI should do it and pass the `encryptionKMSID` in the payload.

Comment (SanjalKatiyar) [reply services/ux-backend/handlers/expandstorage/handler.go:54]: cc @vbnrh

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [reply services/ux-backend/handlers/expandstorage/handler.go]: then enable encryption checkbox should be hidden or disabled then filesystem is selected cc @vbnrh

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:55]: Added the field `encryptionKMSID`

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (travisn) [inline services/ux-backend/handlers/expandstorage/handler.go:175]:
```suggestion
func createCephBlockPoolStorageClass(w http.ResponseWriter, client client.Client, storageClassName, poolName, reclaimPolicy, volumeBindingMode string, enableEncryption bool, encryptionKMSID, namespace string) {
```

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (travisn) [inline services/ux-backend/handlers/expandstorage/handler.go:60]: Today our only scenario might be for LSO, but it's good to keep the naming more generic in case other scenarios are added in the future. If this would be confused with the storage class name we are generating, perhaps we could name it something like `StorageClassForOSDs`.
```suggestion
		StorageClassName string              `json:"storageClassName"`
```

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (travisn) [inline services/ux-backend/handlers/expandstorage/handler.go:153]: Is this feature enabled for stretch clusters? The pool spec is a bit different in that case, where we need to set `replicasPerFailureDomain: 2`, and the `subFailureDomain`

**Review by travisn — CHANGES_REQUESTED** [review]

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (SanjalKatiyar) [reply services/ux-backend/handlers/expandstorage/handler.go:153]: @travisn by any chance Rook takes care of this automatically ??
Asking because UI never sets this on day-2 BlockPools.
Also, I didn't find any reference of it in OCS operator for day-1 BlockPool (unless I missed, @malayparida2000 can u plz confirm ??)

Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:152]: Correct we don't have this enabled for `BlockPools` UI currently, so it is not a common scenario,
We can have this added once we have Blockpool UI using it,

Comment (malayparida2000) [reply services/ux-backend/handlers/expandstorage/handler.go:152]: On the default cephblockpool that ocs-operator creates the Size & the Replica per failure domain field are set accordingly if it's a stretch cluster. https://github.com/red-hat-storage/ocs-operator/blob/c079ab81fd02767ddccc8c24bd324e6c4df89090/controllers/storagecluster/generate.go#L135-L148.

Comment (malayparida2000) [reply services/ux-backend/handlers/expandstorage/handler.go:152]: If extra blockpools are being created via UI those would not be reconciled by ocs operator. So adding the options during creation as Travis suggests seems to be a good idea, but not sure why UI didn't have this configured till now

Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go]: Added a arbiter check

**Comment by parth-gr** [comment]
@rchikatw @travisn @malayparida2000 @SanjalKatiyar @rewantsoni ready for the approval?

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (malayparida2000) [inline services/ux-backend/handlers/expandstorage/handler.go:92]: // create cephFilesystem dataPool

**Review by malayparida2000 — COMMENTED** [review]
mostly LGTM, just these 2 neat suggestions

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (rewantsoni) [inline services/ux-backend/handlers/expandstorage/handler.go:75]: Is this only supported for storagecluster in operator namespace?

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (rewantsoni) [inline services/ux-backend/handlers/expandstorage/handler.go:220]: Since storageClass params cannot be updated after it is created once, would it make sense to use create here? and if the params differ then delete and create it again?

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (rewantsoni) [inline services/ux-backend/handlers/expandstorage/handler.go:293]: Since storageClass params cannot be updated after it is created once, would it make sense to use create here? and if the params differ then delete and create it again?

File: services/ux-backend/main.go
Comment (rewantsoni) [inline services/ux-backend/main.go:83]:
```suggestion
		klog.Exitf("failed to add cephv1 to scheme. %v", err)
```

File: services/ux-backend/main.go
Comment (rewantsoni) [inline services/ux-backend/main.go:86]:
```suggestion
		klog.Exitf("failed to add storagev1 to scheme. %v", err)
```

**Review by rewantsoni — CHANGES_REQUESTED** [review]

**Comment by SanjalKatiyar** [comment]
LGTM.

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:75]: Yes we do run storage cluster on same namespace as per this feature

**Review by malayparida2000 — APPROVED** [review]
/lgtm

**Comment by parth-gr** [comment]
cc @iamniting PTAL

**Review by Madhu-1 — COMMENTED** [review]
It looks like this design is more specific to LSO, shouldnt we try to make the API or code more generic instead of targeting only LSO (ocs-operator/backend server)?

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:47]: As per the design we will use the name of the (LSO storageclass === Device Class)

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:46]: We support adding multiple dataPools for a fileSystem

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:50]: It is common for both, 
Only encryption will be not be used for cephfs

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:60]: We are adding support for per-device-set OSD encryption.
SO some device Sets can have encryption and some dont

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go:95]: SubvolumeGroup currently is not supported, we can look it for provider mode,

We are adding new dataPools on the same filesystem to increase the storage.

**Comment by parth-gr** [comment]
> It looks like this design is more specific to LSO, shouldnt we try to make the API or code more generic instead of targeting only LSO (ocs-operator/backend server)?

Thanks, Removed any specific instance of LSO

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (parth-gr) [reply services/ux-backend/handlers/expandstorage/handler.go]: Yes currently the feature is only for LSO storageclass

**Comment by parth-gr** [comment]
@Madhu-1 @iamniting @rewantsoni can we get a final approval, 
We would  like to test the end-to-end changes with UI using the build,  https://github.com/red-hat-storage/odf-console/pull/1662, 
Currently, we are just able to test it by manually editing csv's and images

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (iamniting) [inline services/ux-backend/handlers/expandstorage/handler.go:137]: Why do we need a `CreateOrUpdate`? I believe just update is enough.

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (iamniting) [inline services/ux-backend/handlers/expandstorage/handler.go:260]: Can you pls correct this err msg

**Review by iamniting — CHANGES_REQUESTED** [review]
Can you pls separate manual changes and generated changes?

Also, if the pool and storage class already exist, should we not send a response indicating that the request has already been fulfilled or something similar? So that the requester/UI already knows that things are in place already.

**Comment by parth-gr** [comment]
> Can you pls separate manual changes and generated changes?

Added a separate commit

> Also, if the pool and storage class already exist, should we not send a response indicating that the request has already been fulfilled or something similar? So that the requester/UI already knows that things are in place already.

This endpoint is only called when there is new LSO storageclass is used, that means we always need to create the pools and storageclass
PLus UI currently checks the ok status with 200 status code

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (iamniting) [inline services/ux-backend/handlers/expandstorage/handler.go:248]: Pls fix this msg.

**Review by iamniting — CHANGES_REQUESTED** [review]
Can you have below 3 files in first commit and others in second commit

`controllers/storagecluster/storageclasses.go`
`services/ux-backend/handlers/expandstorage/handler.go`
`services/ux-backend/main.go`

**Comment by iamniting** [comment]
> > Can you pls separate manual changes and generated changes?
> 
> Added a separate commit
> 
> > Also, if the pool and storage class already exist, should we not send a response indicating that the request has already been fulfilled or something similar? So that the requester/UI already knows that things are in place already.
> 
> This endpoint is only called when there is new LSO storageclass is used, that means we always need to create the pools and storageclass PLus UI currently checks the ok status with 200 status code

The names are passed by the UI. If the same names exist we should send succeed msg IMO. But let's hear from @SanjalKatiyar

**Comment by SanjalKatiyar** [comment]
> The names are passed by the UI. If the same names exist we should send succeed msg IMO. But let's hear from @SanjalKatiyar

UI has validations in place so this should not happen for happy-path, but in case it happens IMO it's okay to show error.
K8s API server also returns error if we try to create an existing CR (same name/namespace). Also, it can happen that specs/parameters of already existing pool/class are different than what user wants now in new pool/class, so better to show error and let user use new name instead.

**Comment by parth-gr** [comment]
> > The names are passed by the UI. If the same names exist we should send succeed msg IMO. But let's hear from @SanjalKatiyar
> 
> UI has validations in place so this should not happen for happy-path, but in case it happens IMO it's okay to show error. K8s API server also returns error if we try to create an existing CR (same name/namespace). Also, it can happen that specs/parameters of already existing pool/class are different than what user wants now in new pool/class, so better to show error and let user use new name instead.

As UI is sending a unique name, I don't this so we need to re-validate,
It's a single endpoint call which already loaded with three/four API operation,
We should not overload it.

If the blockpool wasn't created anyway user will see the error in the UI.

File: services/ux-backend/handlers/expandstorage/handler.go
Comment (iamniting) [reply services/ux-backend/handlers/expandstorage/handler.go:248]: We sorted it over a call.

**Pull request merged**
